### PR TITLE
Backup Cognito user pool and copy it to S3 

### LINKED ISSUE #1: how to preserve user list across cognito user pool rebuilds
If we have to recreate our user pool (e.g. we modify some attributes and Terraform decides to tear it down and rebuild it), we lose our user list.  This is inconvenient, because then people have to recreate their user accounts.  Further, we suspect that upon account recreation, users will get a different user ID than they had before, meaning that the concatenation of the `iss` and `sub` fields in the JWT, used for authentication via group ACLs, will change.  This would be very inconvenient, because it would essentially make worthless all the WebIDs listed in our ACLs, which would essentially cut off editing access for everyone.

So, two questions:
1. How hard is it to regularly back up our user list, and how hard is it to rebuild the same list of users upon user pool rebuild?  I suspect the latter is pretty feasible, since users can be created programmatically using the AWS SDK (though maybe not super-quick since we're new to cognito).  But I've not looked into the backup issue at all, so no idea there.
    1. not super important, but i think we can avoid forcing the user to re-confirm their account upon recreation (see https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/admin-create-user.html, which lists a `CONFIRMED` user status option).
2. Can we re-create the user so they end up with the same `sub`?  If not, we're going to have to come up with a way to re-map everyone's old WebID to their new one in the ACLs whenever the pool is rebuilt.  That sounds like a non-trivial amount of programming work, with a small amount of engineering design and UI design (likely CLI like the other ACL stuff).    If that's to be done manually... well, that sounds like a really awful data entry job, and also super error prone.

### PR DESCRIPTION
**Pull request by mjgiarlo**
Includes:

* Run `npm init` and add `jest` and `eslint`
* Set up CircleCI, CodeClimate, jest, and eslint and update the README
* Dockerize app and correct typo in README
* Add bin/backup stub along with supporting source stub and add babel-node as a dependency
* Add config, dotenv, and babel packages and make sure linter, tests, and bin script all function
* Export user data from Cognito
* Copy user data to S3
* :100: test coverage

Fixes #1

### CODE REVIEW

**Review by jermnelson — APPROVED** [review]
LGTM; @mjgiarlo I'll be happy to merge when this is a real PR 😃

File: .babelrc
Comment (mjgiarlo) [inline .babelrc:3]: Needed to make babel and dotenv play nicely together.

File: .babelrc
Comment (mjgiarlo) [inline .babelrc:10]: Needed to use `async`/`await` in test suite.

File: __mocks__/cognito-fake.js
Comment (mjgiarlo) [inline __mocks__/cognito-fake.js:5]: Yes, this is gross, but this is the interface of the dependency we're faking so `/shrug`!

File: __mocks__/cognito-pagination-fake.js
Comment (mjgiarlo) [inline __mocks__/cognito-pagination-fake.js:7]: This makes sure we can split a result set into paginated chunks without endless recursion.

File: __tests__/CopyUsers.test.js
Comment (mjgiarlo) [inline __tests__/CopyUsers.test.js:43]: The key value includes an ISO representation of a timestamp, so without gnarly date/time mocking, this is the best I could do here.

File: index.js
Comment (mjgiarlo) [inline index.js:1]: This module is untested because it's using code that's already covered in our test suite.

File: package.json
Comment (mjgiarlo) [inline package.json:9]: I don't love the env vars here either, but it works and I wasn't sure there was a better approach. Explained here: https://github.com/LD4P/sinopia_user_backup/pull/2/files#diff-5a61b7aef07cd51690e0d92206685355R13

File: src/CopyUsers.js
Comment (mjgiarlo) [inline src/CopyUsers.js]: Store users in a JSON file named after the user pool in a subdirectory that is the datestamp of the operation.

File: src/configureAWS.js
Comment (mjgiarlo) [inline src/configureAWS.js:1]: This function lives outside the other classes to reduce copypasta.

File: __mocks__/cognito-fake.js
Comment (ndushay) [reply __mocks__/cognito-fake.js:5]: I wouldn't mind having your PR comment be in the actual code ...

File: __mocks__/cognito-pagination-fake.js
Comment (ndushay) [reply __mocks__/cognito-pagination-fake.js:7]: ditto on putting this comment in the code

File: __tests__/CopyUsers.test.js
Comment (ndushay) [reply __tests__/CopyUsers.test.js:43]: I'd put this one in the code as well.

File: src/CopyUsers.js
Comment (ndushay) [reply src/CopyUsers.js]: wouldn't we want the directory named for the userPool, and the filename being the dataestamp?  Possibly even preceded by "users", so 234zzv9wer9ads8f/usersTIMESTAMP.json  ?   then I'd have all the files for the same userpool together and just look for the last one.

Also ... would we want to delete older files?

**Review by ndushay — CHANGES_REQUESTED** [review]
One gen-uine question and a few quibbles.

File: src/CopyUsers.js
Comment (jmartin-sul) [reply src/CopyUsers.js]: > wouldn't we want the directory named for the userPool, and the filename being the dataestamp?

one possible argument for organizing by date before pool id is that it makes it easier to see things in chronological order by directory structure and file name, using lexical sorting in either `tree` or `ls`.  which would make it easy to see at a glance where the latest backup file was, without having to remember what order the pool IDs were used in.

>  Possibly even preceded by "users", so 234zzv9wer9ads8f/usersTIMESTAMP.json

i do like the idea of a filename that's more descriptive than just a pool id or a timestamp, and i do like having files that sort well as siblings even when their parent directory structures get flattened.  but i also like the idea of timestamp preceding pool id.  so my personal pref would be something like `DATE/user-backup_DATE_POOLID.json`, for e.g. a filename like `2019-05-01/user-backup_2019-05-01_234zzv9wer9ads8f.json`.

> Also ... would we want to delete older files?

i'm not sure if this is captured in a ticket at the moment, but my understanding from slack discussion yesterday was that we'd punt on cleanup till later.  i think @mjgiarlo's back of the envelope calculation expected < 25 MB of storage use per year if we don't purge anything, which seems pretty minimal.

Comment (mjgiarlo) [reply src/CopyUsers.js]: @ndushay :speech_balloon: 
> wouldn't we want the directory named for the userPool, and the filename being the datestamp?

Perhaps. But AFAIK we only have one user pool per env, so the top-level user pool dir is only so valuable in an env that won't have multiple pools. I've just created a poll in Slack to see what folks want.

> Possibly even preceded by "users", so 234zzv9wer9ads8f/usersTIMESTAMP.json ? then I'd have all the files for the same userpool together and just look for the last one.

Hmmm, how do you think the `users` prefix helps?

> Also ... would we want to delete older files?

I defer to @jermnelson. Jeremy, how much do we care about cleaning up our user pool backups in M3 (or even in M4)? It's fairly cheap to keep the files hanging around, and we can always clean later. I'm inclined to keep that out of this PR, but I'm happy to write up a new issue if this is something we care about!

Comment (jermnelson) [reply src/CopyUsers.js]: I don't think we need to delete the files at least for a couple of months into the project. To me this would be ongoing maintenance or part of the next work-cycle if we wanted to add a programmatic way to remove the old files.

Comment (mjgiarlo) [reply src/CopyUsers.js]: @jermnelson OK. I'll write up a ticket and it'll sit in the backlog.

Comment (mjgiarlo) [reply src/CopyUsers.js]: https://github.com/LD4P/sinopia_user_backup/issues/3

**Review by ndushay — APPROVED** [review]

**Pull request merged**
